app/main/register-ledger-listeners.ts

```diff
@@ -0,0 +1,169 @@
+import { ipcMain } from 'electron';
+import { BehaviorSubject, Subject, timer, combineLatest, from, of } from 'rxjs';
+import { delay, filter, map, switchMap, take } from 'rxjs/operators';
+import type Transport from '@ledgerhq/hw-transport';
+import TransportNodeHid from '@ledgerhq/hw-transport-node-hid';
+import { safeAwait } from '@blockstack/ui';
+import StacksApp, {
+  LedgerError,
+  ResponseAddress,
+  ResponseSign,
+  ResponseVersion,
+} from '@zondax/ledger-blockstack';
+
+const POLL_LEDGER_INTERVAL = 1_250;
+const SAFE_ASSUME_REAL_DEVICE_DISCONNECT_TIME = 2_000;
+const STX_DERIVATION_PATH = `m/44'/5757'/0'/0/0`;
+
+type LedgerEvents =
+  | 'create-listener'
+  | 'remove-listener'
+  | 'waiting-transport'
+  | 'disconnected'
+  | 'has-transport';
+
+type LedgerStateEvents =
+  | Record<'name', LedgerEvents>
+  | (Record<'name', keyof typeof LedgerError> & ResponseVersion & { action: 'get-version' });
+
+interface LedgerMessageType {
+  type: 'ledger-event';
+}
+
+export type LedgerMessageEvents = LedgerStateEvents & LedgerMessageType;
+
+const ledgerState$ = new Subject<LedgerStateEvents>();
+
+const listeningForDevice$ = new BehaviorSubject(false);
+ipcMain.on('create-ledger-listener', () => listeningForDevice$.next(true));
+ipcMain.on('remove-ledger-listener', () => listeningForDevice$.next(false));
+
+const ledgerDeviceBusy$ = new BehaviorSubject(false);
+
+const transport$ = new BehaviorSubject<Transport | null>(null);
+
+let subscription: null | ReturnType<typeof TransportNodeHid.listen> = null;
+function createDeviceListener() {
+  if (subscription) subscription.unsubscribe();
+  subscription = TransportNodeHid.listen({
+    next: async event => {
+      if (event.type === 'add') {
+        ledgerState$.next({ name: 'waiting-transport' });
+        if (subscription) subscription.unsubscribe();
+        const [error, ledgerTransport] = await safeAwait(TransportNodeHid.open(event.descriptor));
+        ledgerState$.next({ name: 'has-transport' });
+        if (ledgerTransport) transport$.next(ledgerTransport);
+        if (error) console.log({ error });
+      }
+    },
+    error: e => console.log('err', e),
+    complete: () => console.log('complete'),
+  });
+}
+
+export function registerLedgerListeners(webContent: Electron.WebContents) {
+  ledgerState$
+    .pipe(map(ledgerEvent => ({ type: 'ledger-event', ...ledgerEvent })))
+    .subscribe(event => webContent.send('message-event', event));
+}
+
+listeningForDevice$.subscribe(listening => {
```

Comment:
This file is pretty big, could you comment some of what this does? I am not really familiar with rxjs, and as such it seems strange to me that we have these function calls just within a file, rather than in a component or hook or another function.

Reply:
Agree, I've done my best to tidy up. Though important to note this that this is main thread, so running in node.

I've moved the ledger operations into their own folder. Everything else is closely tied to mgmt of the Ledger connection.